src/Features/Core/Portable/DocumentSymbols/IDocumentSymbolsService.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.DocumentSymbols
+{
+    internal interface IDocumentSymbolsService : ILanguageService
+    {
+        /// <summary>
+        /// Gets the symbols defined in this document.
+        /// </summary>
+        Task<ImmutableArray<DocumentSymbolInfo>> GetSymbolsInDocumentAsync(Document document, DocumentSymbolsOptions options, CancellationToken cancellationToken);
+    }
+
+    internal enum DocumentSymbolsOptions
+    {
+        /// <summary>
+        /// Computes all types and nested types as part of the main array returned by
+        /// <see cref="IDocumentSymbolsService.GetSymbolsInDocumentAsync(Document, DocumentSymbolsOptions, CancellationToken)"/>,
+        /// with their top-level members returned as the <see cref="DocumentSymbolInfo.ChildrenSymbols"/> of those types.
+        /// No info on locals or local functions is returned. Members not in the given document can be returned. This format
+        /// matches the behavior of the Visual Studio toolbars.
+        /// </summary>
+        TypesAndMembersOnly,
+        /// <summary>
+        /// Computes all types as a hierarchy. Nested types are returned as <see cref="DocumentSymbolInfo.ChildrenSymbols"/> of their
+        /// containing types. Information about locals and local functions is also returned. Members not in the given document are not
+        /// returned.
+        /// </summary>
+        FullHierarchy
+    }
+
+    /// <summary>
+    /// Information about the symbols in this document
+    /// </summary>
+    internal sealed class DocumentSymbolInfo
```

Comment:
that this is not a record makes me sad.

Reply:
I might be able to make this a record. I had originally intended to make this a struct, but when I did I got very weird test failures that `DocumentSymbolInfo` wasn't able to be loaded by the test assembly.